core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java

```diff
@@ -99,7 +102,143 @@
           "org.apache.calcite.adapter.jdbc.JdbcRules$");
 
   public RelJson(@Nullable JsonBuilder jsonBuilder) {
+    this(jsonBuilder, RelJson::inputTranslatorImpl);
+  }
+
+  private static RexNode inputTranslatorImpl(
+      Map<String, Object> stringObjectMap,
+      RexBuilder rexBuilder,
+      List<RelNode> relNodes) {
+    final Integer input = (Integer) stringObjectMap.get("input");
+    if (input != null) {
+      int i = input;
+      for (RelNode inputNode : relNodes) {
+        final RelDataType rowType = inputNode.getRowType();
+        if (i < rowType.getFieldCount()) {
+          final RelDataTypeField field = rowType.getFieldList().get(i);
+          return rexBuilder.makeInputRef(field.getType(), input);
+        }
+        i -= rowType.getFieldCount();
+      }
+      throw new RuntimeException("input field " + input + " is out of range");
+    } else {
+      throw new RuntimeException("input not defined");
+    }
+  }
+
+  public RelJson(@Nullable JsonBuilder jsonBuilder, InputTranslator inputTranslator) {
     this.jsonBuilder = jsonBuilder;
+    this.inputTranslator = inputTranslator;
+  }
+
+  /**
+   * Transforms a RexNode tree defined in a map (from a JSON) into a RexNode,
+   * applying a special method to inputs instead of transforming them into inputRef.
+   * @param cluster The optimization environment
+   * @param apply is a InputTranslator lambda that transforms the map representing input
+   *               references into a RexNode
+   * @param o the map derived from a RexNode transformed into a JSON
+   * @return the transformed RexNode
+   */
+  public static RexNode readExpression(
```

Comment:
Let's move this method a bit further down the file. It's not the most important method in this class.

The description 'Transforms a RexNode tree' is a bit misleading, given that the input is JSON not RexNode. Maybe something like 'Converts a JSON expression...'?

Convert the anonymous inner class that implements RelInput into a private static inner class. That gives us an opportunity to document it. Also it will make it easier to refactor in future.